### ISSUE TITLE
"Convert match to let-else" assist should handle non-trivial patterns

### ISSUE THREAD

**Issue by jonas-schievink** [rust-lang/rust-analyzer#13540]
Currently, for code like

```rust
let (x, y) = match Some((0, 1)) {
    Some(it) => it,
    None => todo!(),
};
```

The "Convert match to let-else" assist doesn't trigger, because it only works with identifier patterns, but here we have `(x, y)` as the pattern. This is a pretty arbitrary limitation, and the assist should be able to turn it into:

```rust
let Some((x,y)) = Some((0, 1)) else { todo!() };
```

**Comment by jonas-schievink** [comment]
It should also handle more complex expressions in the diverging branch (preceding non-diverging expressions for example)

**Comment by liamhession** [comment]
Hi @jonas-schievink - i'm hoping to try to contribute to this. After familiarizing myself with the contribution guide for rust-analyzer.... Would this issue involve solely making changes within https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-assists/src/handlers/convert_match_to_let_else.rs ?

Going out on a limb - the work involved would be to expand the scope of what the assist looks for in the AST to trigger it, so that a let statement like you identified would cause it to trigger. And then adding some tests? Cheers

**Comment by jonas-schievink** [comment]
Yeah, that's right
